File: src/main/kotlin/com/ywc/spark/kafka/serialization/AvroSerializer.java
package com.ywc.spark.kafka.serialization;

import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.specific.SpecificDatumWriter;
import org.apache.avro.specific.SpecificRecordBase;
import org.apache.kafka.common.serialization.Serializer;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Map;

/**
 * @author yanweichen
 * @date 2018/6/19
 */
public class AvroSerializer implements Serializer<SpecificRecordBase> {
    @Override
    public void configure(Map<String, ?> configs, boolean isKey) {

    }

    @Override
    public byte[] serialize(String topic, SpecificRecordBase data) {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        BinaryEncoder binaryEncoder = EncoderFactory.get().directBinaryEncoder(byteArrayOutputStream
                , null);
        try {
            new SpecificDatumWriter<SpecificRecordBase>(data.getSchema()).write(data, binaryEncoder);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return byteArrayOutputStream.toByteArray();
    }

    @Override
    public void close() {

    }
}

File: src/main/kotlin/com/ywc/spark/kafka/serialization/AvroSerializer.kt
package com.ywc.spark.kafka.serialization

import org.apache.avro.specific.SpecificRecordBase
import org.apache.kafka.common.serialization.Serializer
import org.apache.avro.io.EncoderFactory
import org.apache.avro.specific.SpecificDatumWriter
import java.io.ByteArrayOutputStream


class AvroSerializer<T : SpecificRecordBase> : Serializer<T> {

    override fun configure(configs: MutableMap<String, *>?, isKey: Boolean) {
    }

    override fun serialize(topic: String?, data: T): ByteArray = ByteArrayOutputStream().run {
        EncoderFactory.get().directBinaryEncoder(this, null).run {
            SpecificDatumWriter<T>(data.schema).write(data, this)
        }
        this.toByteArray()
    }

    override fun close() {
    }
}